packages/nimble-components/src/table-column/enum-base/index.ts

```diff
@@ -0,0 +1,123 @@
+import {
+    attr,
+    Notifier,
+    Observable,
+    observable,
+    Subscriber
+} from '@microsoft/fast-element';
+import type {
+    TableStringField,
+    TableBooleanField,
+    TableNumberField
+} from '../../table/types';
+import { TableColumn } from '../base';
+import { Mapping } from '../../mapping/base';
+import type { MappingKeyType } from './types';
+import type { MappingConfig } from './models/mapping-config';
+import type { MappingKey } from '../../mapping/base/types';
+import { resolveKeyWithType } from './models/mapping-key-resolver';
+
+export type TableColumnEnumCellRecord =
+    | TableStringField<'value'>
+    | TableBooleanField<'value'>
+    | TableNumberField<'value'>;
+export type MappingConfigs = Map<MappingKey, MappingConfig>;
+export interface TableColumnEnumColumnConfig {
+    mappingConfigs: MappingConfigs;
+}
+
+/**
+ * Base class for table columns that map values to content (e.g. nimble-table-column-enum-text and nimble-table-column-icon)
+ */
+export abstract class TableColumnEnumBase<
+    TColumnConfig extends TableColumnEnumColumnConfig,
+    TEnumValidator
+>
+    extends TableColumn<TColumnConfig>
+    implements Subscriber {
+    // To ensure the validator is available when other properties get initialized
+    // (which can trigger validation), declare the validator first.
+    /** @internal */
+    public validator = this.createValidator();
+
+    /** @internal */
+    public mappingNotifiers: Notifier[] = [];
+
+    /** @internal */
+    @observable
+    public mappings: Mapping[] = [];
+
+    @attr({ attribute: 'field-name' })
+    public fieldName?: string;
+
+    @attr({ attribute: 'key-type' })
+    public keyType: MappingKeyType = 'string';
+
+    /**
+     * @internal
+     *
+     * Triggers a request to update the columnConfig when any observable property on
+     * a mapping is updated.
+     */
+    public handleChange(source: unknown, args: unknown): void {
+        if (source instanceof Mapping && typeof args === 'string') {
+            this.updateColumnConfig();
+        }
+    }
+
+    public abstract createValidator(): TEnumValidator;
+
+    /**
+     * Called when any Mapping related state has changed
+     */
+    protected abstract updateColumnConfig(): void;
+
+    // Assumes the mapping element state is validated
```

Comment:
I don't see what that would improve. There would still be an assumption that validation had run. Checking if the validation state _isn't_ valid would not catch the case where validation hadn't been run. In a sense, `createColumnConfig` is double-checking that the state is valid (by checking `key !== undefined`) and doing something appropriate if it is not.  In `TableColumnEnumText.createMappingConfig()` we do also handle the case where our configuration is invalid by throwing an error. That's because we assume it would be a programming error, but maybe if we are loosening the assumption that validity has been checked, we can just return a `MappingTextConfig` with an empty string. Thoughts?

Reply:
I think there are a few things that are a bit confusing to me, and perhaps they can be addressed with minimal code changes.
- Is the assumption that _validation has run_? Or, is the assumption that _the column & its mappings are valid_? Can the language in the comments be more precise?
- The comment above `createMappingConfig` is confusing to me because it says that the function "assumes the mapping element state is validated", but there isn't an implementation associated with the function. Should the comment instead say something to the effect of "An implementation of this function may assume that the `mapping` passed in is valid"? If so, why are we doing a check for an allowed mapping type in our implementation? Should performing a check really be the guidance -- "throw an Error from this function if an invalid `Mapping` is passed"?
- It seems like the comments above the functions are intended to be documentation for the functions rather than comments associated with the code. If that's the case, can you switch them to using `/** */` style comments that document the function and also make note of any assumptions associated with the function?
- If every mapping column implementation is expected to check the validity of the configuration before calling `createColumnConfig()`, why not put the validity check there? I agree it doesn't guarantee that the caller has updated their validity, but I think that would still be better than having identical logic in every column implementation for `this.validator.isValid() ? this.createColumnConfig() : undefined`

Reply:
1. I pushed the `this.validator.isValid` check down into `createColumnConfig`, like you suggested.
2. I attempted to run the validation (using a new abstract method) in the base class (i.e. in `updateColumnConfig`), but a typing incompatibility makes that impractical. Instead, I've just updated the comments to be more directly instructive.